### PR TITLE
Go/cassandra client pool fix

### PR DESCRIPTION
**Pull request by gokcanozakdag**
**Goals (and why)**:
Adressing the issue: https://github.com/palantir/atlasdb/issues/2973
**Implementation Description (bullets)**:
`CassandraClient.rawClient()` method is removed, and all the methods used from `rawClient` is wrapped in `CassandraClient`. `CassandraClient` alters the state of `isValid` if a blacklisted exception is thrown by any method calls. 
**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:
`CassandraClient` may be the right place to start.
**Priority (whenever / two weeks / yesterday)**:

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3063)

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java [hunk 1]
```diff
@@ -21,27 +21,32 @@
...
 import org.apache.cassandra.thrift.UnavailableException;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TProtocol;
 
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 
 @SuppressWarnings({"all"}) // thrift variable names.
 public interface CassandraClient {
-    Cassandra.Client rawClient();
+    boolean isValid();
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java]: A javadoc might be helpful here.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java [hunk 2]
```diff
@@ -81,4 +86,41 @@ CqlResult execute_cql3_query(CqlQuery cqlQuery,
             ConsistencyLevel consistency)
             throws InvalidRequestException, UnavailableException, TimedOutException, SchemaDisagreementException,
             org.apache.thrift.TException;
+
+    TProtocol getOutputProtocol();
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java]: nit: ordering seems somewhat random here

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java [hunk 3]
```diff
@@ -20,41 +20,57 @@
...
+import org.apache.thrift.transport.TTransportException;
 
+import com.google.common.collect.ImmutableSet;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 import com.palantir.atlasdb.keyvalue.impl.AbstractKeyValueService;
 
 @SuppressWarnings({"all"}) // thrift variable names.
 public class CassandraClientImpl implements CassandraClient {
     private final Cassandra.Client client;
+    private boolean isValid;
```
[the remark this answers is not in the document]
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java]: Yes, this should be volatile.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java [hunk 4]
```diff
@@ -20,41 +20,57 @@
...
+
+    private static final Set<Class> blackListedExceptions =
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java]: style: `BLACKLISTED_EXCEPTIONS` (casing for constants, and blacklist is typically a single word). Did you run checkstyle?

Again, would be nice to write in a comment what it means for these exceptions to be blacklisted. Does it mean that we do consider the client invalid if this exception is hit, or that we don't?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java [hunk 5]
```diff
@@ -124,10 +233,50 @@ public CqlResult execute_cql3_query(CqlQuery cqlQuery,
...
     private ColumnParent getColumnParent(TableReference tableRef) {
         return new ColumnParent(AbstractKeyValueService.internalTableName(tableRef));
     }
+
+    private <T> T executeMethod(ThrowingSupplier<T> supplier)
+            throws InvalidRequestException, UnavailableException, TimedOutException, SchemaDisagreementException, TException {
+        try {
+            return supplier.apply();
+        } catch (Exception e) {
+            isValid = blackListedExceptions.stream()
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java]: I don't agree with assigning `isValid` here. An individual client can go from the state valid -> invalid, but should NOT go from invalid to valid.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java [hunk 6]
```diff
@@ -124,10 +233,50 @@ public CqlResult execute_cql3_query(CqlQuery cqlQuery,
...
+            throw e;
+        }
+    }
+
+    private void executeMethod(ThrowingVoidSupplier supplier)
+            throws InvalidRequestException, UnavailableException, TimedOutException, SchemaDisagreementException, TException {
+        try {
+            supplier.apply();
+        } catch (Exception e) {
+            isValid = !blackListedExceptions.stream()
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java]: Looks like either this or the above is wrong. Again, is a blacklisted exception one that we ignore, or the opposite?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java [hunk 7]
```diff
@@ -124,10 +233,50 @@ public CqlResult execute_cql3_query(CqlQuery cqlQuery,
...
+                    .anyMatch(b -> b.isInstance(e));
+            throw e;
+        }
+    }
+
+    private <T> T executeMethodWithoutException(Supplier<T> supplier) {
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java]: This is awkward :( not sure I see a way around it, if overloading the method causes compiler confusion.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java [hunk 8]
```diff
@@ -124,10 +233,50 @@ public CqlResult execute_cql3_query(CqlQuery cqlQuery,
...
 
-        return client.execute_cql3_query(queryBytes, compression, consistency);
+        return executeMethod(() -> client.execute_cql3_query(queryBytes, compression, consistency));
     }
 
     private ColumnParent getColumnParent(TableReference tableRef) {
         return new ColumnParent(AbstractKeyValueService.internalTableName(tableRef));
     }
+
+    private <T> T executeMethod(ThrowingSupplier<T> supplier)
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java]: `executeMethod` is pretty vague. `executeHandlingExceptions`?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java [hunk 9]
```diff
@@ -124,7 +124,12 @@ private CassandraClient instrumentClient(Client rawClient) {
         }
     }
 
-    static Cassandra.Client getClientInternal(InetSocketAddress addr, CassandraKeyValueServiceConfig config)
+    static CassandraClient getClientInternal(InetSocketAddress addr, CassandraKeyValueServiceConfig config)
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java]: Can we now make this private?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java [hunk 10]
```diff
@@ -124,10 +233,50 @@ public CqlResult execute_cql3_query(CqlQuery cqlQuery,
...
+            throws InvalidRequestException, UnavailableException, TimedOutException, SchemaDisagreementException, TException {
+        try {
+            return supplier.apply();
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientImpl.java]: For safety, do we want to check `isValid` before executing any method?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java (hunk 9 above)
Comment (gokcanozakdag) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java]: It is still used by other classes :(

**Pull request merged**